storage/fcds/fcds.go

```diff
@@ -0,0 +1,360 @@
+	if !s.shardHasFreeOffsets(shard) {
+		// shard does not have free offset
+		return -1, false, nil
+	}
+
+	offset = -1 // negative offset denotes no available free offset
+	if s.freeCache != nil {
+		// check if local cache has an offset
+		offset = s.freeCache.get(shard)
+	}
+
+	if offset < 0 {
+		// free cache did not return a free offset,
+		// check the meta store for one
+		offset, err = s.meta.FreeOffset(shard)
+		if err != nil {
+			return 0, false, err
+		}
+	}
+	if offset < 0 {
+		// meta store did not return a free offset,
+		// mark this shard that has no free offsets
+		s.markShardWithFreeOffsets(shard, false)
+		return -1, false, nil
+	}
+
+	return offset, true, nil
+}
+
+// Delete removes chunk data.
+func (s *Store) Delete(addr chunk.Address) (err error) {
+	done, err := s.protect()
+	if err != nil {
+		return err
+	}
+	defer done()
+
+	shard := getShard(addr)
+	s.markShardWithFreeOffsets(shard, true)
+
+	mu := s.shards[shard].mu
+	mu.Lock()
+	defer mu.Unlock()
+
+	if s.freeCache != nil {
+		m, err := s.getMeta(addr)
+		if err != nil {
+			return err
+		}
+		s.freeCache.set(shard, m.Offset)
+	}
+	return s.meta.Remove(addr, shard)
+}
+
+// Count returns a number of stored chunks.
+func (s *Store) Count() (count int, err error) {
+	return s.meta.Count()
+}
+
+// Iterate iterates over stored chunks in no particular order.
+func (s *Store) Iterate(fn func(chunk.Chunk) (stop bool, err error)) (err error) {
+	done, err := s.protect()
+	if err != nil {
+		return err
+	}
+	defer done()
+
+	for _, sh := range s.shards {
+		sh.mu.Lock()
+	}
+	defer func() {
+		for _, sh := range s.shards {
+			sh.mu.Unlock()
+		}
+	}()
+
+	return s.meta.Iterate(func(addr chunk.Address, m *Meta) (stop bool, err error) {
+		data := make([]byte, m.Size)
+		_, err = s.shards[getShard(addr)].f.ReadAt(data, m.Offset)
+		if err != nil {
+			return true, err
+		}
+		return fn(chunk.NewChunk(addr, data))
+	})
+}
+
+// Close disables of further operations on the Store.
+// Every call to its methods will return ErrDBClosed error.
+// Close will wait for all running operations to finish before
+// closing its MetaStore and returning.
+func (s *Store) Close() (err error) {
+	s.quitOnce.Do(func() {
+		close(s.quit)
+	})
+
+	done := make(chan struct{})
+	go func() {
+		s.wg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(15 * time.Second):
+	}
+
+	for _, sh := range s.shards {
+		if err := sh.f.Close(); err != nil {
+			return err
+		}
+	}
+	return s.meta.Close()
+}
+
+// protect protects Store from executing operations
+// after the Close method is called and makes sure
+// that Close method will wait for all ongoing operations
+// to finish before returning. Returned function done
+// must be closed to unblock the Close method call.
+func (s *Store) protect() (done func(), err error) {
+	select {
+	case <-s.quit:
+		// Store is closed.
+		return nil, ErrDBClosed
+	default:
+	}
+	s.wg.Add(1)
+	return s.wg.Done, nil
```

Comment:
Why not have a private `Store.done()` function instead of returning `done` function?

Reply:
The main reason is to make it more obvious that `done` must be called in order to finish the protection and that it is only in the relation to protect function. But for this small codebase, I think that it is not so significant. Is it ok to call the new method `unprotect` instead `done`?

Reply:
Both make sense to me, I agree with your point as well, so go for what feels better for you. If you decide to make it a private function `unprotect` sounds good to me.

Reply:
The usage is nicer with a private method, with one less line. Thanks.